homeassistant/helpers/update_coordinator.py

```diff
@@ -82,11 +82,13 @@ def __init__(
         self._debounced_refresh = request_refresh_debouncer
 
     @callback
-    def async_add_listener(self, update_callback: CALLBACK_TYPE) -> Callable[[], None]:
+    def async_add_listener(
+        self, update_callback: CALLBACK_TYPE, context: object | None = None
+    ) -> Callable[[], None]:
         """Listen for data updates."""
         schedule_refresh = not self._listeners
 
-        self._listeners.append(update_callback)
+        self._listeners[update_callback] = context
```

Comment:
This might be a bit problematic. The current code seem to handle passing the same update_callback many times, then i think you can call remove listener same amount of times to completely remove the the callback. It's now assumed to be unique.

Reply:
I've switched to using the "return cleanup" approach for the coordinator, that resolves this ambiguity and allow multiple calls on same update_callback (theoretically we could pass along the context, but that would be a rather large change for many components)